admin-dev/themes/new-theme/js/pages/product/edit/product-partial-updater.js

```diff
@@ -0,0 +1,160 @@
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+const {$} = window;
+
+/**
+ * When product is edited we want to send only partial updates
+ * so this class compares the initial data from the form computes
+ * the diff when form is submitted And dynamically build another
+ * form to submit only updated data (along with required fields
+ * token and such).
+ *
+ * It also disabled the submit button as long as no data has been
+ * modified by the user.
+ */
+export default class ProductPartialUpdater {
+  constructor($productForm, $productFormSubmitButton) {
+    this.$productForm = $productForm;
+    this.$productFormSubmitButton = $productFormSubmitButton;
+  }
+
+  /**
+   * This the public method you need to use to start this component
+   * ex: new ProductPartialUpdater($productForm, $productFormSubmitButton).watch();
+   */
+  watch() {
+    this.$productFormSubmitButton.prop('disabled', true);
+    this.initialData = this.getFormDataAsObject();
+    this.$productForm.submit((e) => this.updatePartialForm(e));
+    this.$productForm.on('change', ':input', () => this.updateSubmitButtonState());
+  }
+
+  /**
+   * This methods handles the form submit
+   *
+   * @param event
+   * @returns {boolean}
+   * @private
+   */
+  updatePartialForm(event) {
+    event.stopImmediatePropagation();
+
+    const updatedData = this.getUpdatedFormData();
+    if (updatedData !== null) {
+      this.postUpdatedData(updatedData);
+    } else {
+      // @todo: This is temporary we should probably use a nice modal instead, that said since the submit button is
+      //        disabled when no data has been modified it should never happen
+      alert('no fields updated');
+    }
+
+    return false;
+  }
+
+  /**
+   * Dynamically build a form with provided updated data and submit this "shadow" form
```

Comment:
Do we have to build this shadow form?
Is it done so we can benefit from form-related features on backend side with Symfony Validations for example?

Which brings an interesting question:
- true form is filled
- we build the shadow form from properties which were changed
- shadow form submitted
- one input has validation issue
How do we get the validation error message and display it on true form? Do we refresh the whole page?

Reply:
When the form is submitted (wether it's shadow or not) the page is refreshed Since the shadow form is built with the same form names Symfony doesn't know (and doesn't care) which form sent the dat it just matches the expected format thanks to form names

And for the same reason if an input has validation errors then when the page is rendered after form invalidation then the appropriate input has the error attached to it And the best part is it only happens on modified inputs since the other ones are ignored, which is intended for partial update